src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -155,17 +155,18 @@ public function dump(array $options = array())
         }
 
         (new AnalyzeServiceReferencesPass(false, !$this->getProxyDumper() instanceof NullDumper))->process($this->container);
+        $checkedNodes = array();
         $this->circularReferences = array();
-        foreach (array(true, false) as $byConstructor) {
-            foreach ($this->container->getCompiler()->getServiceReferenceGraph()->getNodes() as $id => $node) {
-                if (!$node->getValue() instanceof Definition) {
-                    continue;
-                }
-                $currentPath = array($id => true);
-                $this->analyzeCircularReferences($node->getOutEdges(), $currentPath, $id, $byConstructor);
+        foreach ($this->container->getCompiler()->getServiceReferenceGraph()->getNodes() as $id => $node) {
+            if (!$node->getValue() instanceof Definition) {
+                continue;
+            }
+            if (!isset($checkedNodes[$id])) {
+                $this->analyzeCircularReferences($id, $node->getOutEdges(), $checkedNodes);
             }
         }
         $this->container->getCompiler()->getServiceReferenceGraph()->clear();
+        $checkedNodes = array();
```

Comment:
looks unused?  should it be `unset($checkedNodes);`?

Reply:
could be but that would provide anything more so not really needed

Reply:
unset would avoid IDE or code style checker marking it as unused variable so nobody would remove it accidentally ;).

But I'm happy that the container now again only take some seconds to build 😄